modules/navigation_mesh_generator/register_types.cpp

```diff
@@ -0,0 +1,62 @@
+/*************************************************************************/
+/*  register_types.cpp                                                   */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "register_types.h"
+
+#include "core/config/engine.h"
+#include "navigation_mesh_generator.h"
+
+#ifdef TOOLS_ENABLED
+#include "editor/navigation_mesh_editor_plugin.h"
+#endif // TOOLS_ENABLED
+
+NavigationMeshGenerator *_nav_mesh_generator = nullptr;
+
+void initialize_navigation_mesh_generator_module(ModuleInitializationLevel p_level) {
+	if (p_level == MODULE_INITIALIZATION_LEVEL_SERVERS) {
+		_nav_mesh_generator = memnew(NavigationMeshGenerator);
+		GDREGISTER_CLASS(NavigationMeshGenerator);
+		Engine::get_singleton()->add_singleton(Engine::Singleton("NavigationMeshGenerator", NavigationMeshGenerator::get_singleton(), "NavigationMeshGenerator"));
```

Comment:
I think the `NavigationMeshGenerator` singleton should remain in `servers` as an abstract class.  Then, this module should implement that singleton with Recast.

Like the `NavigationServer`, we'd have a manager class that allows setting the mesh generator that Godot will use.  This module would then set itself to be that implementation.

The nav mesh editor plugin also doesn't seem to be tied to Recast and could probably be in core Godot and apply to all implementations of the generator.

This would open the door to other implementations of the mesh generator, perhaps for games that use a different navigation system that isn't based on finding walkable regions in 3D geometry.

Reply:
So you mean basically turning it into a full server for navmesh baking.

I did not want to go as far with this pr but I noticed a lot of dependency and init chain issues with the geometry parsers from the other pr and turning it into a server with a module override in the background would solve many of them. Might be beneficial should 2D baking and real threading support be added at some point to already have a server singleton available to extend. Sigh ... you convinced me, off I go to a lot of work.